test/e2e/reconciler/namespace/controller_test.go

```diff
@@ -157,20 +157,51 @@ func TestNamespaceScheduler(t *testing.T) {
 					GenerateName: "e2e-nss-2-",
 				},
 				Spec: clusterv1alpha1.ClusterSpec{
-					KubeConfig: string(server2Kubeconfig),
+					KubeConfig:    string(server2Kubeconfig),
+					Unschedulable: true, // cordoned.
 				},
 			}, metav1.CreateOptions{})
 			require.NoError(t, err, "failed to create cluster2")
 
-			t.Log("Delete the old cluster and expect the namespace to end up scheduled to the new cluster.")
+			t.Log("Delete the old cluster and expect the namespace to end up unschedulable (since the new cluster is cordoned).")
 
 			err = clusterClient.Delete(ctx, cluster1.Name, metav1.DeleteOptions{})
 			require.NoError(t, err, "failed to delete cluster1 %q", cluster1.Name)
 
+			err = expect(namespace, unschedulableMatcher())
+			require.NoError(t, err, "did not see namespace marked unschedulable")
+
+			t.Log("Uncordon the new cluster and expect the namespace to end up scheduled to it.")
+
+			cluster2.Spec.Unschedulable = false // uncordoned.
+			cluster2, err = clusterClient.Update(ctx, cluster2, metav1.UpdateOptions{})
+			require.NoError(t, err, "failed to uncordon cluster2")
+
+			err = expect(namespace, scheduledMatcher(cluster2.Name))
+			require.NoError(t, err, "did not see namespace marked scheduled for cluster2 %q", cluster2.Name)
+
+			t.Log("Mark the new cluster for immediate eviction, and see the namespace get unschedulable.")
+
+			now := metav1.NewTime(time.Now().Add(-time.Hour))
+			cluster2, err = clusterClient.Get(ctx, cluster2.Name, metav1.GetOptions{})
+			require.NoError(t, err, "failed to get cluster2")
+			cluster2.Spec.EvictAfter = &now
+			cluster2, err = clusterClient.Update(ctx, cluster2, metav1.UpdateOptions{})
+			require.NoError(t, err, "failed to evict cluster2")
+
+			err = expect(namespace, scheduledMatcher(cluster2.Name))
```

Comment:
Should the namespace remain scheduled to cluster2?

Reply:
Good catch. The matcher was passing presumably because of delay in unscheduling it. Fixed.